build.rs

```diff
@@ -15,79 +31,26 @@ fn main() {
     }
-
-    // https://rust-lang-nursery.github.io/rust-bindgen
-    // https://docs.rs/bindgen
-    let mut builder = bindgen::Builder::default()
-        // This is needed if use `#include <nng.h>` instead of `#include "path/nng.h"` in wrapper.h
-        //.clang_arg("-Inng/src/")
-        .header("wrapper.h");
-
-    if !cfg!(feature = "legacy-111-rc4") {
-        builder = builder
-            .whitelist_type("nng_.*")
-            .whitelist_function("nng_.*")
-            .whitelist_var("NNG_.*")
-            // Generate `pub const NNG_UNIT_EVENTS` instead of `nng_unit_enum_NNG_UNIT_EVENTS`
-            .prepend_enum_name(false)
-            // Generate `pub enum ...` instead of multiple `pub const ...`
-            .rustified_enum("nng_.*_enum")
-            // Enum special cases:
-            .rustified_enum("nng_pipe_ev")
-            .rustified_enum("nng_sockaddr_family")
-            .rustified_enum("nng_zt_status")
-            // no_std support
-            // https://rust-embedded.github.io/book/interoperability/c-with-rust.html#automatically-generating-the-interface
-            .ctypes_prefix("cty")
-            .use_core();
+    #[cfg(not(feature = "build-nng"))]
+    {
+        println!("cargo:rustc-link-lib=dylib=nng");
     }
-    let bindings = builder.generate().expect("Unable to generate bindings");
-    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
-    bindings
-        .write_to_file(out_path.join("bindings.rs"))
-        .expect("Couldn't write bindings");
 }
 
-fn cmake() -> PathBuf {
-    // Compile-time features
-    let generator = if cfg!(feature = "cmake-ninja") {
-        "Ninja"
-    } else if cfg!(feature = "cmake-vs2017") {
-        "Visual Studio 15 2017"
-    } else if cfg!(feature = "cmake-vs2017-win64") {
-        "Visual Studio 15 2017 Win64"
-    } else {
-        // Default generators
-        if cfg!(target_family = "unix") {
-            "Unix Makefiles"
-        } else {
-            "Ninja"
-        }
-    };
-    let stats = if cfg!(feature = "nng-stats") {
-        "ON"
-    } else {
-        "OFF"
-    };
-    let tls = if cfg!(feature = "nng-tls") {
-        "ON"
-    } else {
-        "OFF"
-    };
+fn generator() -> &'static str {
+    let ninja = cfg!(feature = "cmake-ninja");
+    let vs2017 = cfg!(feature = "cmake-vs2017");
+    let vs2017_win64 = cfg!(feature = "cmake-vs2017-win64");
+    let unix = cfg!(target_family = "unix");
 
-    // Run cmake to build nng
-    Config::new("nng")
-        .generator(generator)
-        .define("CMAKE_BUILD_TYPE", "Release")
-        .define("NNG_TESTS", "OFF")
-        .define("NNG_TOOLS", "OFF")
-        .define("NNG_ENABLE_STATS", stats)
-        .define("NNG_ENABLE_TLS", tls)
-        .build()
+    match (ninja, vs2017, vs2017_win64, unix) {
+        (true, false, false, _) => "Ninja",
+        (false, true, false, _) => "Visual Studio 15 2017",
+        (false, false, true, _) => "Visual Studio 15 2017 Win64",
```

Comment:
These permit `(false, false, true, true)`, for example.
This is kind of unwieldy.  Off the top of my head it's simpler to do:
```rust
if ninja as i32 + vs2017 as i32 .... > 1 {
    panic!();
}
```
Surprised cargo features doesn't support options that are mutually exclusive....

Reply:
`(false, false, true, true)` arguably doesn't really matter - that last field is there only to provide a default for when no generator is specified. But I can definitely see the argument for failing early, as the VS + Unix combination is very unlikely to work.

I think adding the flags is less intuitive and clear, but the build script is so simple that it honestly doesn't mater to me.

I'm also surprised by how Cargo features seem lackluster. Something like mutual exclusivity seems like it would be simple to implement and do 90% of the things people have developed hacks for.